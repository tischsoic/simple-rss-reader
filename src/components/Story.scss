h5 {
  margin-bottom: 0;
  margin-top: 0;
}

.story img {
  float: left;

  height: 150px;
  width: 150px;
}

.story p {
  float: right;
}

.story,
.story div {
  background-color: lightgrey;
  margin-bottom: 10px;
  margin-top: 10px;
  padding: 10px;
  overflow: auto;
}

.read-story,
.read-story div {
  background-color: gray;
}

button.favorite {
  float: right;
  // position: absolute;
  // top: 0;
  // right: 0;
}
