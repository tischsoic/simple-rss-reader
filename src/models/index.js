import { Record, List } from 'immutable';

export const FeedRecord = Record({
  link: null,   // link is an ID
  title: null,
});

export const StoryRecord = Record({
  id: null,
  feedId: null,
  rssId: null,

  title: null,
  description: null,
  link: null,
  imageUrl: null,
  publicationDate: null,

  read: false,
  favorite: false,
});

export const AppStateRecord = Record({
  feeds: List(),
  stories: List(),
  favoriteStoriesIds: List(),
  lastId: -1,
});
