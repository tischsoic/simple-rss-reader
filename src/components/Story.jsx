import PropTypes from 'prop-types';
import React from 'react';
import { StoryRecord } from '../models';

import './Story.scss';

const Story = ({
  storyRecord,
  onStoryClick,
  opened,
  addFavoriteStory,
  favoriteList,
}) => {
  const storyImageUrl = storyRecord.get('imageUrl');
  const storyRecordTitle = storyRecord.get('title');
  const storyAlreadyRead = storyRecord.get('read');
  const storyId = storyRecord.get('id');
  const isFavoriteStory = storyRecord.get('favorite');
  const showAddToFavoritesButton = !isFavoriteStory && !favoriteList;

  return (
    <div
      onClick={(event) => {
        event.stopPropagation();
        onStoryClick(storyId);
      }}
      role="presentation"
      className={storyAlreadyRead ? 'read-story story' : 'story'}
    >
      {showAddToFavoritesButton ? (
        <button
          className="favorite"
          onClick={(event) => {
            event.stopPropagation();
            addFavoriteStory(storyId);
          }}
        >
          {'<3'}
        </button>
      ) : null}

      <h5>{storyRecordTitle}</h5>

      {opened ? (
        <div>
          {storyImageUrl ? (
            <img src={storyImageUrl} alt={storyRecordTitle} />
          ) : null}{' '}
          {storyRecord.get('description')}
          <br />
          <a href={storyRecord.get('link')} target="_blank">
            Open article
          </a>
        </div>
      ) : null}
    </div>
  );
};

Story.propTypes = {
  storyRecord: PropTypes.instanceOf(StoryRecord).isRequired,
  onStoryClick: PropTypes.func.isRequired,
  opened: PropTypes.bool.isRequired,
  addFavoriteStory: PropTypes.func.isRequired,
  favoriteList: PropTypes.bool.isRequired,
};

export default Story;
