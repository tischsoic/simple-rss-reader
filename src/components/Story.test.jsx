import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Story from './Story';
import { StoryRecord } from '../models';

describe('Story component', () => {
  const storyRecord = new StoryRecord({
    id: 12,
    feedId: 'feeed-link',
    rssId: 'rssId',

    title: 'story title',
    description: 'story description',
    link: 'story link',
    imageUrl: 'image url',
    publicationDate: Date.parse('Fri, 20 Apr 2018 17:14 EDT'),

    read: false,
    favorite: false,
  });
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Story
          storyRecord={storyRecord}
          opened
          onStoryClick={() => {}}
          addFavoriteStory={() => {}}
          favoriteList={false}
        />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('calls addFavoriteStory when button is clicked', () => {
    const addFavoriteStoryMock = jest.fn();
    const rectangle = shallow(
      <Story
        storyRecord={storyRecord}
        opened
        onStoryClick={() => {}}
        addFavoriteStory={addFavoriteStoryMock}
        favoriteList={false}
      />,
    );

    rectangle.find('button').simulate('click', {
      stopPropagation: () => {},
    });
    expect(addFavoriteStoryMock.mock.calls.length).toEqual(1);
    expect(addFavoriteStoryMock.mock.calls[0][0]).toEqual(
      storyRecord.get('id'),
    );
  });
});
