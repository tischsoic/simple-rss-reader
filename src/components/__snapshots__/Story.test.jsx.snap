// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Story component renders correctly 1`] = `
<div
  className="story"
  onClick={[Function]}
  role="presentation"
>
  <button
    className="favorite"
    onClick={[Function]}
  >
    &lt;3
  </button>
  <h5>
    story title
  </h5>
  <div>
    <img
      alt="story title"
      src="image url"
    />
     
    story description
    <br />
    <a
      href="story link"
      target="_blank"
    >
      Open article
    </a>
  </div>
</div>
`;
